engine/src/execution_graph/logic_controllers/BatchOrderByProcessing.cpp

```diff
@@ -525,39 +556,45 @@ LimitKernel::LimitKernel(std::size_t kernel_id, const std::string & queryString,
-
-        eventTimer.start();
-        std::tie(input, rows_limit) = ral::operators::limit_table(std::move(input), rows_limit);
-        eventTimer.stop();
-
-        auto log_output_num_rows = input->num_rows();
-        auto log_output_num_bytes = input->sizeInBytes();
-
-        events_logger->info("{ral_id}|{query_id}|{kernel_id}|{input_num_rows}|{input_num_bytes}|{output_num_rows}|{output_num_bytes}|{event_type}|{timestamp_begin}|{timestamp_end}",
-                        "ral_id"_a=context->getNodeIndex(ral::communication::CommunicationData::getInstance().getSelfNode()),
-                        "query_id"_a=context->getContextToken(),
-                        "kernel_id"_a=this->get_id(),
-                        "input_num_rows"_a=log_input_num_rows,
-                        "input_num_bytes"_a=log_input_num_bytes,
-                        "output_num_rows"_a=log_output_num_rows,
-                        "output_num_bytes"_a=log_output_num_bytes,
-                        "event_type"_a="compute",
-                        "timestamp_begin"_a=eventTimer.start_time(),
-                        "timestamp_end"_a=eventTimer.end_time());
-
-        output->addToCache(std::move(input));
+    try{
+        CodeTimer eventTimer(false);
+        auto & input = inputs[0];
+        if (rows_limit<0) {
+            output->addToCache(std::move(input));
+        } else {
+            auto log_input_num_rows = input->num_rows();
+            auto log_input_num_bytes = input->sizeInBytes();
+
+            eventTimer.start();
+            std::tie(input, rows_limit) = ral::operators::limit_table(std::move(input), rows_limit);
+            eventTimer.stop();
+
+            auto log_output_num_rows = input->num_rows();
+            auto log_output_num_bytes = input->sizeInBytes();
+
+            events_logger->info("{ral_id}|{query_id}|{kernel_id}|{input_num_rows}|{input_num_bytes}|{output_num_rows}|{output_num_bytes}|{event_type}|{timestamp_begin}|{timestamp_end}",
+                            "ral_id"_a=context->getNodeIndex(ral::communication::CommunicationData::getInstance().getSelfNode()),
+                            "query_id"_a=context->getContextToken(),
+                            "kernel_id"_a=this->get_id(),
+                            "input_num_rows"_a=log_input_num_rows,
+                            "input_num_bytes"_a=log_input_num_bytes,
+                            "output_num_rows"_a=log_output_num_rows,
+                            "output_num_bytes"_a=log_output_num_bytes,
+                            "event_type"_a="compute",
+                            "timestamp_begin"_a=eventTimer.start_time(),
+                            "timestamp_end"_a=eventTimer.end_time());
+
+            output->addToCache(std::move(input));
+        }
+    }catch(rmm::bad_alloc e){
+        return {ral::execution::task_status::RETRY, std::string(e.what()), std::vector< std::unique_ptr<ral::frame::BlazingTable> > ()};
```

Comment:
this is a particularly interesting kernel. It can modify the inputs BUT applying limit_table more than once should NOT have negative consequences. Meaning that if we make limit_table some how take a const reference to input, keeping input in scope because it stays in the inputs vector, we could actually still be able to recover from any error except for output->addToCache()

Reply:
So, i do think we want to have `ral::operators::limit_table` take a const reference, so that we still have control of input. But `ral::operators::limit_table` may also have wanted to return `input` untouched. So i think we would want to do something like this:

```
std::unique_ptr<ral::frame::BlazingTable> limited_input;
bool output_is_just_input;
std::tie(limited_input, output_is_just_input, rows_limit) = ral::operators::limit_table(input, rows_limit);
```
Where if you have output_is_just_input == true, then limited_input is nullptr and you are supposed to just use input.

Then you also move `output->addToCache(std::move(input));` to outside the try catch, so that you can move the input into the RETRY.
And also, it would be something like this:

```
if (output_is_just_input)
     output->addToCache(std::move(input));
else
     output->addToCache(std::move(limited_input));
````

Reply:
Fixed.